### PR TITLE
Add haloperidol, potassium iodide

### PR DESCRIPTION
**Pull request by slarticodefast**
## About the PR
This PR adds haloperidol and potassium iodide from SS13 (see the [tg station wiki](https://tgstation13.org/wiki/Guide_to_chemistry)). The drowsiness and radiation protection status effects have been implemented. Some existing entity and reagent prototypes have been adjusted to use them as well.

## Why / Balance
Originally I only wanted to add the extradimensional orange to botany, but it needed haloperidol as a reagent which in turn requires potassium iodide. So I went down the rabbit hole and implemented both.

Potassium iodide gives you a radiation protection status effect at a rate of 4 seconds per unit. Inspired by the [real-life chemical](https://en.wikipedia.org/wiki/Potassium_iodide), its purpose is to serve as a short term alternative to the radiation suit. I gave it 90% radiation protection, since I think full protection would be too powerful and you should not be able to walk around right next to the singularity. To prevent you from chugging a whole jug of it, I added an overdose of 20u. I also added potassium iodine pills to the radiation treatment kit, replacing the emergency medipen.

The effect of haloperidol is the same as in SS13. It removes most stimulating/hallucinogenic drugs from the body, reduces their effects and makes you drowsy. It could be used for patient treatment, especially by the psychologist or as a weapon for security to deal with drugged up maniacs. The SS13 wiki mentions it being used in tranquilizer darts. The strenght of removing the rainbow effect has been set be lower than that of synaptizine, since the latter also poisons you.

Drowsiness works similar to drunkenness and narcolepsy. It blurs your vision around the edges and makes you fall asleep in random intervals. The reagents that make you drowsy now also make you yawn to indicate the status effect setting in. The effect disappears after a while or can be reduced by drinking coffee, cola or taking stimulating drugs. The sleep interval is random from 5 to 60 seconds (compared to 300-600 for narcolepsy) with a duration of 10 to 30 seconds (the same as narcolepsy).

I decided to change chloral hydrate to make you drowsy instead of forcing you to sleep instantly above a certain threshold. This is now more similar to its SS13 counterpart and distuingishes it from nocturine. At small doses, where the status effect is shorter than 60 seconds, there is a chance you dont fall asleep at all. I read stories about the feared CMO armed with a chloral hydrate hypopen as an instant knockout weapon. The random aspect in the drowsy effect should give this some balance. Nocturine has been unchanged.

## Technical details
Potassium idodide works by adding a server side status effect component and its corresponding system. The sytem applies a new DamageModifierSet with 90% radiation protection upon taking damage.

Drowsiness works by adding a shared status effect component, similar to that of the DrunkComponent. The server side deals with making the player fall asleep. The code is mostly identical to that of the narcolepsy system. I have thought about adding the narcoplepsy component instead with different variables, but this would require a lot of additional logic to deal with players who are both drowsy and narcoleptic, which might cause drowsiness to cure narcolepsy, change its incident times, etc. On the client side it adds a new overlay and a new shader, which both work similar to the DrunkOverlay or the RainbowOverlay. The shader generates a simple radial blur with adjustable strength.

I am not very familiar with the ECS, so I hope my implementation is not too hacky.
Things I would like feedback on include:
- Balancing
- Should drowsiness also slow you down or should it be kept separate?
- Should drowsiness add an alert icon? Realistically it is something you would be aware of.
- Improving the shader performance. It currently uses highp floats, which maybe could be changed to lowp. I am not sure about the performance impact.
- Are there additional reagents/drinks that should add or remove the new status effects? Especially in soda.yml I think some of the softdrinks probably contain caffeine. For now only cola and coffee have been adjusted to make you less drowsy.

## Media
potassium iodide:

https://github.com/space-wizards/space-station-14/assets/161409025/23bd8311-fe9d-46ba-9ce8-7203eb92f20e

haloperidol:

https://github.com/space-wizards/space-station-14/assets/161409025/283e1172-7fb9-4bb3-90a1-d860d05d83e6

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
none

**Changelog**
:cl:
- add: Added potassium iodide. It gives you short term radiation protection and can be found in radiation treatment kits.
- add: Added haloperidol. It removes most stimulating/hallucinogenic drugs from the body and makes you drowsy.
- add: Added the drowsiness status effect. It blurs your vision and makes you randomly fall asleep. Drink some coffee or cola to remove it.
- tweak: Chloral hydrate now makes you drowsy instead of forcing you to sleep instantly.

### CODE REVIEW

**Comment by Cojoke-dot** [comment]
hm... since you are making drowsiness for the sleeping chems, should you not apply it to no2 and noct too?

**Comment by slarticodefast** [comment]
> hm... since you are making drowsiness for the sleeping chems, should you not apply it to no2 and noct too?

NO2 and nocturine need the forced sleep effect to be reliable tools. You don't want patients to wake up during surgery and nocturine should stay an instant knockout weapon. Maybe not replace it, but add it? That way you are drowsy for a while after you wake up from the forced sleep effect.

**Comment by dffdff2423** [comment]
Maybe rename Haloperidol to something else because it is the name of an IRL drug? It could be a bit too close to home for someone.

**Comment by slarticodefast** [comment]
> Maybe rename Haloperidol to something else because it is the name of an IRL drug? It could be a bit too close to home for someone.

Good point. I just took the name from SS13 and didn't even know it was a real thing. Any name suggestions? Maybe something close to the word "reality" or something to do with the extradimensional orange?

**Comment by Hrosts** [comment]
> Maybe rename Haloperidol to something else because it is the name of an IRL drug? It could be a bit too close to home for someone.

Hypnodoline?

**Comment by Bugfinder1** [comment]
> Maybe rename Haloperidol to something else because it is the name of an IRL drug? It could be a bit too close to home for someone.

Thats not a good reason to change the name in my opinion.

Counter-Arguments: 
1 - We have a handful of Meds named after their IRL counterpart, and it never has been an issue.
2 - People migrating over from SS13 or who have slight medical knowledge can guess what the med does by its name.

**Comment by deltanedas** [comment]
haloperidol in fish game is an anti psychotic drug so it could also cause confusion

**Comment by Golinth** [comment]
> Maybe rename Haloperidol to something else because it is the name of an IRL drug? It could be a bit too close to home for someone.

Diphenhydramine, Epinephrine, Ipecac, Saline, Tranexamic Acid, Diphenylmethylamine, Mannitol, and probably a couple of others that I missed. Having IRL drug names hasn't been a problem before, why should it be a problem now? And if negative experiences are a factor, Diphenhydramine stands out as an incredibly common (but ineffective) way to attempt suicide, and that didnt stop it.

**Comment by slarticodefast** [comment]
> Its almost being used the same here. Anti-Hallucinogenic. When psychology has actual content, I could see it being used for something very similar, treating schizophrenia or other breakdowns. 

That is the plan. In SS13 it removes hallucinations. Right now the only hallucination we have is SeeingRainbows, but I am thinking about making a design doc for adding more.

> Thats not a good reason to change the name in my opinion.

I see pychosis as a very sensitive topic and understand why someone would like the name to be changed. On the other hand SS14 has many very adult themes and got an age restriction for a reason. I would leave the decision whether the name should be changed for a maintainer as I didn't find any guidelines for this.

File: Resources/Textures/Shaders/drowsiness.swsl
Comment (mirrorcult) [inline Resources/Textures/Shaders/drowsiness.swsl:6]: would prefer to see this file just called `radial-blur.swsl` or something then, seems reusable

File: Resources/Prototypes/Shaders/shaders.yml
Comment (mirrorcult) [inline Resources/Prototypes/Shaders/shaders.yml:63]: same here

File: Resources/Locale/en-US/reagents/meta/medicine.ftl
Comment (mirrorcult) [inline Resources/Locale/en-US/reagents/meta/medicine.ftl:146]:
```suggestion
reagent-desc-haloperidol = Removes most stimulating and hallucinogenic drugs. Reduces druggy effects and jitteriness. Causes drowsiness.
```

File: Content.Server/Radiation/Components/RadiationProtectionComponent.cs
Comment (mirrorcult) [inline Content.Server/Radiation/Components/RadiationProtectionComponent.cs:14]: think this could very easily be genericized to just like a `DamageProtectionBuffComponent`, its not really specific to radiation except that the default field is one for rads

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Resources/Textures/Shaders/drowsiness.swsl
Comment (slarticodefast) [reply Resources/Textures/Shaders/drowsiness.swsl]: done

File: Resources/Prototypes/Shaders/shaders.yml
Comment (slarticodefast) [reply Resources/Prototypes/Shaders/shaders.yml]: done

File: Resources/Locale/en-US/reagents/meta/medicine.ftl
Comment (slarticodefast) [reply Resources/Locale/en-US/reagents/meta/medicine.ftl]: done

File: Content.Server/Radiation/Components/RadiationProtectionComponent.cs
Comment (slarticodefast) [reply Content.Server/Radiation/Components/RadiationProtectionComponent.cs:17]: I like the idea but the problem here is its use as a status effect. StatusEffectsSystem.cs adds this component and removes it after a certain time is up. For example, if we had an radiation protection status effect and a seperate electrocution protection status effect, the component for both would get removed when the time for one is up. So right now each status effect requires its own component.

**Comment by slarticodefast** [comment]
Unrelated integration tests fail? One seems to be ActionBibleSummon and the other DeleteAllThenGhost.

File: Content.Server/Radiation/Components/RadiationProtectionComponent.cs
Comment (slarticodefast) [reply Content.Server/Radiation/Components/RadiationProtectionComponent.cs:17]: I was trying to create DamageProtectionBuffComponent and make RadiationProtectionComponent just inherit from it, overwrite the ModifierSetId and have a separate component for the status effect that way, but it seems like component inheritance is not a thing in this codebase? The status effect system might need some changes to prevent component spam.

**Comment by Emisse** [comment]
Closed due to feature freeze May 10th-June 14th. Comment to have it reopen after this.

**Comment by slarticodefast** [comment]
Please reopen and set this to draft.
I would like to check for merge conflicts and balance changes before merging.

**Comment by slarticodefast** [comment]
@mirrorcult I addressed all the review points and solved the merge conflict. Also did some minor balance tweaks and added the drowsiness removal to the new energy drinks. Ready for re-review.

**Review by deathride58 — CHANGES_REQUESTED** [review]
Reviewing specifically the shader code here; it could use a cleanup!

**Comment by deathride58** [comment]
In addition to our review, to address one of the parts of the original post:
> Improving the shader performance. It currently uses highp floats, which maybe could be changed to lowp. I am not sure about the performance impact.
Given that SS14's client isn't running on phones (yet; you can never say never with that type of thing), there isn't much difference between the precisions beyond the actual precision, as desktop GPUs are *really* good at handling 32-bit floats (with lower precisions being emulated if the hardware isn't there).
The main thing that's expensive about the performance for the shader in this PR is the `for()` loop contained within. However, *how* expensive this is depends entirely on the GPU and the drivers installed! GPUs are normally *really* bad at branching code like that. Some drivers are decent enough at detecting when a `for()` loop can be reasonably replaced with repetitions of the code within them, but most simply don't bother.
That being said though, with only 16 samples, it's probably fine for most GPUs (it doesn't really hold a candle to some of the more expensive shaders currently in the codebase). If it ends up problematic for iGPU users, then it could be reduced to 8 samples without sacrificing quality *too* much.

**Comment by BolloTea** [comment]
Sorry, most likely this will be outside of this PR, but it would be nice if this drowsiness effect could be added to alcohol, for example, as in case of an overdose, so that a person has a certain chance of falling asleep.  This would add a lot of roleplaying interactions and situations

File: Resources/Textures/Shaders/radial_blur.swsl
Comment (slarticodefast) [reply Resources/Textures/Shaders/radial_blur.swsl]: A leftover from testing I totally forgot to remove.

File: Resources/Textures/Shaders/radial_blur.swsl
Comment (slarticodefast) [reply Resources/Textures/Shaders/radial_blur.swsl]: Seems to work, thanks!

File: Resources/Textures/Shaders/radial_blur.swsl
Comment (slarticodefast) [reply Resources/Textures/Shaders/radial_blur.swsl]: Moved this part to the C# side.

File: Resources/Textures/Shaders/radial_blur.swsl
Comment (slarticodefast) [reply Resources/Textures/Shaders/radial_blur.swsl:8]: Good point, but I thought a little about it and decided to leave it as it is. Instead of restricting vision in a circle the idea this shader is to be psychologically close to the feeling of being drowsy by blurring your view "around the edges". So even if you have a widescreen resolution, the effect should not be stronger on the left or right side than on the top or bottom.

**Comment by slarticodefast** [comment]
@deathride58 Thanks a lot for your review. This is my first time working with shaders, so it is highly appreciated. I addressed all of your comments.

> That being said though, with only 16 samples, it's probably fine for most GPUs

It works on my extremely shitty notebook without a performance impact. Just to be sure I reduced the sample count to 10, which still looks good enough.
It actually does not have to be a multiple of two anymore, that comment was from a previous version where the samples were weighted differently.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Shared/Drowsiness/DrowsinessComponent.cs
Comment (slarticodefast) [reply Content.Shared/Drowsiness/DrowsinessComponent.cs]: Done.
Also changed it to use TimeSpan and compare to CurTime instead of accumulating frametime.

**Review by metalgearsloth — APPROVED** [review]

**Comment by Barrett-S** [comment]
Chloral can now be instant woken up from (1.5 seconds) just by spamming the wake-up action 

Couple this with the fact that it randomly activates over 60 seconds and it's become effectively useless for anything but RP.

**Comment by Barrett-S** [comment]
Amount of Chloral in one's system also does not change sleep duration, 90u of chloral will sleep you once for 1.5 seconds before you just die to OD poison effect. Something is for sure busted with the wakeup action interaction

**Comment by slarticodefast** [comment]
Right now the forced sleep time is random between 2 and 5 seconds and you will fall asleep several times if the drowsiness is long enough. This is similar to slip stun times of 1.5-3 seconds and was chosen to be a serious liability to combat, encouraging you to get rid of the status effect, but not long enough to cause an instant defeat. The exact balance is of course subject to change and can be further adjusted. In prevous dicussions (some of them [here](https://github.com/space-wizards/space-station-14/issues/28659) or [here](https://github.com/space-wizards/space-station-14/pull/29879)) the agreement was that a CMO armed with a chloral hydrate hypopen was way too strong. The restricted nocturine still serves that old purpose for traitors.

**Comment by Barrett-S** [comment]
The most I saw when testing was 3 seconds, but I was getting 2 seconds or less most of the time. I think the interaction with the wake-up action is potentially bugged and overrides the duration.

And to be fair, those previous discussion were largely downvoted and the points made seemed to be largely unpopular. As far as balance goes, I think the effectiveness of CMO with chloral is over-hyped, and if we did want to nerf it, a better change would be a set delay (longer metabolization) and requiring 16u instead of the current 12u.

This would further distance chloral from noct as it would require double the injections. 

With this change, I don't see Chloral ever being used over mutagen or amatoxin, which is a shame since having a non-hurtful combat chem that was essentially locked to CMO was very fitting.

**Comment by Brandon-Huu** [comment]
Nah this is awful.

**Comment by ZombieChibiXD** [comment]
> Nah this is awful.

This have been a long time coming actually, pre-merge freeze. I should've seen it coming but it does make nocturine a bit more exclusive and actually enough of a threat.

**Comment by slarticodefast** [comment]
I will probably make a minor adjustment PR for the balance, we can have some discussion there. And yeah, this PR was up 3 months and noone gave any balance feedback until now.

**Comment by Brandon-Huu** [comment]
> I will probably make a minor adjustment PR for the balance, we can have some discussion there. And yeah, this PR was up 3 months and noone gave any balance feedback until now.

It's impossible to put someone into a straitjacket now :(

**Comment by ZombieChibiXD** [comment]
> een a long time coming actually, pre-merge fre

Offtopic but cuff and then straitjacket usually works for me.

**Comment by Brandon-Huu** [comment]
> > een a long time coming actually, pre-merge fre
> 
> Offtopic but cuff and then straitjacket usually works for me.

How? The cuff and jacket take up the same slot?

**Comment by ZombieChibiXD** [comment]
https://github.com/user-attachments/assets/6a8ba313-45b1-4292-a0f7-41f619202094

A bit cheese, but this is the concept. It's definitely viable by a 2 person thing. Also additionally if they resist it's an execution either way at this point. A proper fix would be straitjacket override cuff since the same mechanism for cuff is to drop item if the person have items in their hand.

**Comment by Brandon-Huu** [comment]
> Space.Station.14.2024-08-06.11-00-42.mp4
> 
> A bit cheese, but this is the concept. It's definitely viable by a 2 person thing. Also additionally if they resist it's an execution either way at this point. A proper fix would be straitjacket override cuff since the same mechanism for cuff is to drop item if the person have items in their hand.

What?

They just stood there?

**Comment by Brandon-Huu** [comment]
Also, what? You put cuffs and strait jacket on?
Isn't that a bug, are they cuffed forever now?

**Comment by ZombieChibiXD** [comment]
You can definitely uncuff, stun, then recuff, and while the progress bar is moving, start giving straitjacket, as the game only checks for the status at the hand on cuff start.

As for bug, I think it's not neccesarily one, as you just need to uncuff them twice, so no they're not cuffed forever.

Offtopic again, but it's totally in the realm of possibility to 4 cuff someone, and keep them cuffed when they use freedom implant. :godo:
Unless metashield is broken by seeing an open uplink, this is extremely cheese, but I can see it as a real thing happening (I've seen cops cuff someone twice for god knows why, what's 4)

**Pull request merged**
